File: ruby/circular-buffer/circular_buffer.rb
class CircularBuffer
  class BufferEmptyException < StandardError; end
  class BufferFullException < StandardError; end

  def initialize(size)
    @size = size
    @elements = []
  end

  def read
    raise BufferEmptyException if elements.empty?

    elements.pop
  end

  def write(value)
    raise BufferFullException if elements.size == size

    elements.unshift(value)
  end

  def write!(value)
    elements.pop if elements.size == size
    
    write(value)
  end

  def clear
    elements.clear
  end

  private
  attr_reader :size, :elements
end

File: julia/circular-buffer/circular-buffer.jl
mutable struct CircularBuffer{T}
    elements::Vector{T}
    capacity::Integer

    CircularBuffer{T}(capacity::Integer) where {T} = new([], capacity)
end

function Base.push!(cb::CircularBuffer, item; overwrite::Bool=false)
    if length(cb.elements) == cb.capacity
        overwrite ? popfirst!(cb.elements) : throw(BoundsError("Cannot write to full buffer"))
    end

    push!(cb.elements, item)
    cb
end

function Base.popfirst!(cb::CircularBuffer)
    isempty(cb.elements) && throw(BoundsError("Cannot pop empty buffer"))
    popfirst!(cb.elements)
end

function Base.empty!(cb::CircularBuffer)
    empty!(cb.elements)
    cb
end
